iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1110 @@
+                        _mqttClientOptionsBuilder.WithProxy(proxyUri.AbsoluteUri);
+                    }
+                }
+            }
 
-            _qosSendPacketToService = settings.PublishToServerQoS;
-            _qosReceivePacketFromService = settings.ReceivingQoS;
+            MqttClientOptionsBuilderTlsParameters tlsParameters = new MqttClientOptionsBuilderTlsParameters();
 
-            // If the CleanSession flag is set to false, C2D messages will be retained across device sessions, i.e. the device
-            // will receive the C2D messages that were sent to it while it was disconnected.
-            // If the CleanSession flag is set to true, the device will receive only those C2D messages that were sent
-            // after it had subscribed to the message topic.
-            _retainMessagesAcrossSessions = !settings.CleanSession;
+            List<X509Certificate> certs = _connectionCredentials.Certificate == null
+                ? new List<X509Certificate>(0)
+                : new List<X509Certificate> { _connectionCredentials.Certificate };
 
-            _webProxy = settings.Proxy;
+            tlsParameters.Certificates = certs;
 
-            if (channelFactory != null)
+            if (_mqttTransportSettings?.RemoteCertificateValidationCallback != null)
             {
-                _channelFactory = channelFactory;
+                tlsParameters.CertificateValidationHandler = CertificateValidationHandler;
             }
-            else
+
+            tlsParameters.UseTls = true;
+            tlsParameters.SslProtocol = _mqttTransportSettings.SslProtocols;
+            _mqttClientOptionsBuilder
+                .WithTls(tlsParameters)
+                .WithProtocolVersion(MQTTnet.Formatter.MqttProtocolVersion.V311) // 3.1.1
+                .WithCleanSession(_mqttTransportSettings.CleanSession)
+                .WithKeepAlivePeriod(_mqttTransportSettings.IdleTimeout);
+
+            if (_mqttTransportSettings.HasWill && _mqttTransportSettings.WillMessage != null)
             {
-                _channelFactory = settings.Protocol switch
+                _mqttClientOptionsBuilder
+                    .WithWillTopic(_deviceToCloudMessagesTopic)
+                    .WithWillPayload(_mqttTransportSettings.WillMessage.Payload);
+
+                if (_mqttTransportSettings.WillMessage.QualityOfService == QualityOfService.AtMostOnce)
                 {
-                    IotHubClientTransportProtocol.Tcp => CreateChannelFactory(connectionCredentials, _additionalClientInformation, settings),
-                    IotHubClientTransportProtocol.WebSocket => CreateWebSocketChannelFactory(connectionCredentials, _additionalClientInformation, settings),
-                    _ => throw new InvalidOperationException($"Unsupported transport setting: {settings.Protocol}"),
-                };
+                    _mqttClientOptionsBuilder.WithWillQualityOfServiceLevel(MqttQualityOfServiceLevel.AtMostOnce);
```

Comment:
I was about to suggest that we use MqttQualityOfServiceLevel directlty in our settings, but I think it is a good choice that we're exposing our own type and not mqtt library's. That way we can save our sdk from intentional breaks if mqtt library makes any breaking changes.